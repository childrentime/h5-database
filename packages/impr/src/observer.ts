import "intersection-observer";

interface ElementInfo {
  el: HTMLElement;
  info: any;
  once: boolean;
  delay: number;
  sendImprMetrics: (trackingInfo: any) => void;
}

const MIN_RATIO_THRESHOLD = 0.75;
const OBSERVER_CONFIGS = {
  rootMargin: "0px",
  threshold: [MIN_RATIO_THRESHOLD],
};

class ImprObserver {
  #uniqid: number;
  #observer: IntersectionObserver | null;
  #entryQuene: Map<number, IntersectionObserverEntry>;
  #elStore: Map<number, ElementInfo>;
  #stashQuene: ((...args: any) => any)[];
  #lastTimer: ReturnType<typeof setTimeout> | null;
  #sendImprMetrics!: (trackingInfo?: any) => void;

  constructor() {
    this.#uniqid = 1;
    this.#observer = null;
    this.#entryQuene = new Map();
    this.#elStore = new Map();
    this.#stashQuene = [];
    this.#lastTimer = null;
    this.init();
  }

  private init = async () => {
    if (typeof window === "undefined") {
      return;
    }

    this.createOberver();
  };

  private createOberver = () => {
    this.#observer = new IntersectionObserver(
      this.observerCallBack,
      OBSERVER_CONFIGS
    );
    // observer没有建立之前 可能已经有注册曝光的元素了
    Array.isArray(this.#stashQuene) &&
      this.#stashQuene.forEach((func) => {
        func && func();
      });
    this.#stashQuene = [];
  };

  observerCallBack: IntersectionObserverCallback = (entries) => {
    //  与屏幕相交时触发回调，分为进入屏幕和移出屏幕
    entries.forEach((entry) => {
      //  使用opts.inIntersection和otps.intersectionRatio >= threshold均可以判断
      //  但是后者在video标签或内部元素小于外部且外部overflow: hidden的情况会失效
      const uniqId = parseInt(
        (entry.target as HTMLElement).dataset.uniqid || "",
        10
      );
      if (!uniqId) {
        return;
      }

      // Guard against show / hide callback, which doesn't respect threshold.
      // https://github.com/w3c/IntersectionObserver/issues/357
      const shouldTrigger =
        entry.isIntersecting && entry.intersectionRatio >= MIN_RATIO_THRESHOLD;

      const { delay = 0 } = this.#elStore.get(uniqId) || {};
      if (shouldTrigger) {
        if (delay === 0) {
          this.log(uniqId);
          return;
        }
        this.#entryQuene.set(uniqId, entry);
      } else {
        if (this.#entryQuene.get(uniqId)) {
          if (entry.time - this.#entryQuene.get(uniqId)!.time >= delay) {
            this.log(uniqId);
          }
          this.#entryQuene.delete(uniqId);
        }
      }
      //  debounce 防止反复触发
      if (this.#lastTimer) {
        clearTimeout(this.#lastTimer);
        this.#lastTimer = null;
      }
      this.#lastTimer = setTimeout(() => {
        const keys = [...this.#entryQuene.keys()];
        keys.forEach((id) => {
          this.log(id);
          this.#entryQuene.delete(id);
        });
      }, delay);
    });
  };

  log = (id: number) => {
    if (!this.#elStore.get(id)) {
      return;
    }
    const { info, once, el } = this.#elStore.get(id) || {};
    if (!el) {
      return;
    }
    once && this.unRegister(el, id);
    this.#sendImprMetrics(info);
  };

  public register = (elementInfo: ElementInfo) => {
    const { el, info, sendImprMetrics } = elementInfo;
    if (!this.#sendImprMetrics) {
      this.#sendImprMetrics = sendImprMetrics;
    }
    if (!el || !info) {
      return;
    }
    const id = this.#uniqid;
    this.#uniqid++;
    el.dataset.uniqid = String(id);
    this.#elStore.set(id, elementInfo);
    if (this.#observer) {
      this.#observer.observe(el);
    } else {
      this.#stashQuene.push(() => {
        this.#observer!.observe(el);
      });
    }
  };

  unRegister = (el: Element, id: number) => {
    if (this.#observer) {
      this.#observer.unobserve(el);
    }
    this.#elStore.delete(id);
    this.#entryQuene.delete(id);
  };
}

export default new ImprObserver();
