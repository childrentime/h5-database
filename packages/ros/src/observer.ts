import { IObservable } from "./types";

export const autorun = (callback: () => void | (() => void)): (() => void) => {
  const observer = new Observer(callback);
  try {
    observer.trigger();
  } catch (error) {
    observer.dispose();
    throw error;
  }

  return observer.dispose.bind(observer);
};

export const reaction = (
  deps: IObservable[],
  callback: () => void | (() => void)
) => {
  const observer = new Observer(callback);
  for (const dep of deps) {
    dep.register(observer);
    observer.observables.add(dep);
  }

  return observer.dispose.bind(observer);
};

export const when = (predicate: () => boolean, callback: () => void) => {
  const dispose = autorun(() => {
    const v = predicate();
    if (v) {
      callback();
      dispose();
    }
  });

  return dispose;
};

/**
 * 执行上下文，收集observer依赖
 */
export let observerContext: Observer | void = void 0;

/**
 * 所有的autorun instance
 */
export const allContext: Set<Observer> = new Set();

export class Observer {
  status: "up-to-date" | "out-of-date";
  observables: Set<IObservable>;
  /**
   * 回调
   */
  callback: () => void | (() => void);
  /**
   * 清理
   */
  #cleanup: (() => void) | void = void 0;
  /**
   * 解除所有订阅
   */
  dispose() {
    this.observables.forEach((observable) => {
      observable.unregister(this);
    });
    allContext.delete(this);
  }

  constructor(callback: () => void | (() => void)) {
    this.callback = callback;
    this.observables = new Set();
    this.status = "up-to-date";
    allContext.add(this);
  }

  trigger() {
    const prevContext = observerContext;
    observerContext = this;
    if (typeof this.#cleanup === "function") {
      this.#cleanup();
    }
    this.#cleanup = this.callback();
    this.status = "up-to-date";
    observerContext = prevContext;
  }

  track(callback: () => void) {
    console.log('track start')
    const prevContext = observerContext;
    observerContext = this;
    callback();
    this.status = "up-to-date";
    observerContext = prevContext;
    console.log('track end')
  }
}
