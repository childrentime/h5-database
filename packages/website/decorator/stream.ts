import { RootStore } from "../store";

export function serialize(_value: any, { name, addInitializer }: any) {
  addInitializer(function (this: RootStore) { 
    if (!this.serializeKeys) {
      this.serializeKeys = new Set();
    }
    this.serializeKeys.add(name);
  });
}

const isServer = typeof window === 'undefined';
export const promiseMap = new Map<string, (...args: any[]) => Promise<any>>();
export const promiseStatusMap = new Map<string, 'settled' | 'pending'>();
export const throwIfUnresolved = (key: string) => {
  const status = promiseStatusMap.get(key)!;
  if(status !== 'settled'){
    throw Promise.resolve("waiting for promise fulfilled....")
  }
}
export function pending(key: string) {
  return (value: (...args: any) => Promise<any>, {addInitializer,name }: ClassMethodDecoratorContext<any, any>) => {
    addInitializer(function () { 
      const bindValue = value.bind(this); 

      if (!isServer) {
        this[name] = bindValue;
        promiseMap.set(key, this[name]);
      }else {
        this[name] = async (args: any) => {
            return bindValue(args).then((res: any) => {
              promiseStatusMap.set(key,'settled')
              return [key, res];
            }).catch((err) => {
              promiseStatusMap.set(key,'settled')
            })
        };
      }
    });    
  }
}
