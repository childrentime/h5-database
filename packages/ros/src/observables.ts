import { Computed } from "./computed";
import { ObservableArray } from "./observableArray";
import { ObservableMap } from "./observableMap";
import { ObservableObject } from "./observableObject";
import { ObservableSet } from "./observableSet";
import { ObservableValue, PrimitiveType } from "./observableValue";
import { Observer } from "./observer";
import { IObservableMode } from "./types";
import { isES6Map, isES6Set, isPlainObject, isProxy } from "./utils";

export const createChildObservable = (
  value: any,
  contexts: Set<Observer | Computed>
) => {
  if (typeof value === "object" && value !== null) {
    if (isProxy(value)) {
      return value;
    }
    if (Array.isArray(value)) {
      return new ObservableArray(value, {
        parentContexts: contexts,
      }).value;
    }
    if (isPlainObject(value)) {
      return new ObservableObject(value, {
        parentContexts: contexts,
      }).value;
    }
    if (isES6Map(value)) {
      return new ObservableMap(value, {
        parentContexts: contexts,
      }).value;
    }
    if (isES6Set(value)) {
      return new ObservableSet(value, {
        parentContexts: contexts,
      }).value;
    }
    return value;
  }
  return value;
};

export const observable = {
  box<T extends PrimitiveType = any>(value: T): ObservableValue<T> {
    return new ObservableValue(value);
  },
  array<T = any>(
    initialValues: T[],
    mode: IObservableMode = "deep"
  ): ObservableArray<T> {
    return new ObservableArray(initialValues, {
      mode,
    });
  },
  map<K = any, V = any>(
    initialValues: Map<K, V>,
    mode: IObservableMode = "deep"
  ): ObservableMap<K, V> {
    return new ObservableMap(initialValues, {
      mode,
    });
  },
  set<T = any>(
    initialValues: Set<T>,
    mode: IObservableMode = "deep"
  ): ObservableSet<T> {
    return new ObservableSet<T>(initialValues, {
      mode,
    });
  },
  object<T extends object = any>(
    props: T,
    mode: IObservableMode = "deep"
  ): ObservableObject<T> {
    return new ObservableObject<T>(props, { mode });
  },
};
