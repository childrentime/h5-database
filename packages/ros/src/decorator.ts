import { IObservable, IObservableMode } from "./types";
import { observable as innerObservable } from "./observables";
import { isServer } from "@mpa-ssr/core-utils";
import { Computed, computed as innerComputed } from "./computed";

export function observable(
  type: "set" | "map" | "object" | "array" | "box" = "box",
  mode: IObservableMode = "deep"
) {
  return (
    { get, set }: ClassAccessorDecoratorTarget<unknown, any>,
    { name, kind }: ClassAccessorDecoratorContext
  ) => {
    if (kind === "accessor") {
      let obs: IObservable;
      const result: ClassAccessorDecoratorResult<any, any> = {
        init(this, value) {
          if (isServer) {
            obs = value;
            return value;
          }
          /**@ts-ignore */
          obs = innerObservable[type](value, mode);
          return obs.value;
        },
        get() {
          if (isServer) {
            return obs;
          }
          return obs.value;
        },
        set(newValue) {
          if (isServer) {
            obs = newValue;
            return;
          }
          obs.value = newValue;
        },
      };

      return result;
    }
  };
}

export function computed(
  value: () => unknown,
  context: ClassGetterDecoratorContext
) {
  const { kind, name } = context;
  if (kind === "getter") {
    let init = false;
    let compute: Computed;
    return function (this: ThisType<unknown>) {
      if (!init) {
        compute = innerComputed(value.bind(this));
        init = true;
      }
      return compute.value as any;
    };
  }

  throw new Error("can not apply decorator computed bisides getter");
}
