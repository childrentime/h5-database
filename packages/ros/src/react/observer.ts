import { __DEV__, isServer, copyStaticProperties } from "@mpa-ssr/core-utils";
import { forwardRef, memo } from "react";
import { useObserver } from "./useObserver";

const hasSymbol = typeof Symbol === "function" && Symbol.for;
const ReactForwardRefSymbol = hasSymbol
  ? Symbol.for("react.forward_ref")
  : typeof forwardRef === "function" &&
    forwardRef((props: any) => null)["$$typeof"];

export function observer<P extends object, R = {}>(
  baseComponent: React.ForwardRefExoticComponent<
    React.PropsWithoutRef<P> & React.RefAttributes<R>
  >
): React.MemoExoticComponent<
  React.ForwardRefExoticComponent<
    React.PropsWithoutRef<P> & React.RefAttributes<R>
  >
>;

export function observer<P extends object>(
  baseComponent: React.FunctionComponent<P>
): React.FunctionComponent<P>;

export function observer<
  C extends React.FunctionComponent<any> | React.ForwardRefRenderFunction<any>
>(baseComponent: C): C;

/**
 * copy from mobx-react
 */
export function observer<P extends object, R>(
  baseComponent:
    | React.ForwardRefRenderFunction<R, P>
    | React.FunctionComponent<P>
    | React.ForwardRefExoticComponent<
        React.PropsWithoutRef<P> & React.RefAttributes<R>
      >
) {
  if (isServer) {
    return baseComponent;
  }
  let render = baseComponent;
  let useForwardRef = false;

  if (
    ReactForwardRefSymbol &&
    /**@ts-ignore */
    baseComponent["$$typeof"] === ReactForwardRefSymbol
  ) {
    useForwardRef = true;
    /**@ts-ignore */
    render = baseComponent["render"];
  }

  let observerComponent = (props: any, ref: React.Ref<R>) => {
    return useObserver(() => render(props, ref));
  };

  (observerComponent as React.FunctionComponent).displayName =
    baseComponent.displayName;
  Object.defineProperty(observerComponent, "name", {
    value: baseComponent.name,
    writable: true,
    configurable: true,
  });

  if ((baseComponent as any).contextTypes) {
    (observerComponent as React.FunctionComponent).contextTypes = (
      baseComponent as any
    ).contextTypes;
  }

  if (useForwardRef) {
    // `forwardRef` must be applied prior `memo`
    // `forwardRef(observer(cmp))` throws:
    // "forwardRef requires a render function but received a `memo` component. Instead of forwardRef(memo(...)), use memo(forwardRef(...))"
    observerComponent = forwardRef(observerComponent);
  }

  observerComponent = memo(observerComponent);

  copyStaticProperties(baseComponent, observerComponent);

  if (__DEV__) {
    Object.defineProperty(observerComponent, "contextTypes", {
      set() {
        throw new Error(
          `[mobx-react-lite] \`${
            this.displayName ||
            this.type?.displayName ||
            this.type?.name ||
            "Component"
          }.contextTypes\` must be set before applying \`observer\`.`
        );
      },
    });
  }

  return observerComponent;
}
