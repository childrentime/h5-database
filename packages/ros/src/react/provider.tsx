import { PropsWithChildren, createContext } from "react";

export type IValueMap = {
  rootStore?: any;
};
export const RosProviderContext = createContext<any>({});

export function RosProvider(props: PropsWithChildren<IValueMap>) {
  const { children, rootStore } = props;

  return (
    <RosProviderContext.Provider value={rootStore}>
      {children}
    </RosProviderContext.Provider>
  );
}


RosProvider.displayName = "RosProvider"