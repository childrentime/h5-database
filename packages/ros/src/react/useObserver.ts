import { isServer } from "@mpa-ssr/core-utils";
import { useCallback, useRef, useState } from "react";
import { Observer } from "../observer";
import { useSyncExternalStore } from "use-sync-external-store/shim";

export function useObserver(render: () => any) {
  if (isServer) {
    return render();
  }

  const storeChangeRef = useRef<() => void>();
  const observerRef = useRef<Observer>();
  const symbolRef = useRef<Symbol>();
  const [, forceUpdate] = useState<Symbol>()

  if(!symbolRef.current){
    symbolRef.current = Symbol();
  }

  if (!observerRef.current) {
    observerRef.current = new Observer(() => {
      symbolRef.current = Symbol();
      storeChangeRef.current?.();
    });
  }

  const subscribe = useCallback((onStoreChange: () => void) => {
    storeChangeRef.current = onStoreChange;
    /**
     *  We've lost our reaction and therefore all subscriptions, occurs when <StrictMode>
     */
    if(!observerRef.current){
      observerRef.current = new Observer(() => {
        symbolRef.current = Symbol();
        storeChangeRef.current?.();
      });
      forceUpdate(Symbol())
    }

    return () => {
      storeChangeRef.current = undefined;
      observerRef.current?.dispose();
      observerRef.current = undefined;
    };
  }, []);

  const getSnapshot = useCallback(() => {
    return symbolRef.current;
  }, []);

  const getServerSnapShot = useCallback(() => {
    return symbolRef.current;
  }, []);

  useSyncExternalStore(subscribe, getSnapshot, getServerSnapShot);

  let renderResult;
  let exception;


  observerRef.current!.track(() => {
    try {
      renderResult = render();
    } catch (error) {
      exception = error;
    }
  });

  if (exception) {
    throw exception; // re-throw any exceptions caught during rendering
  }

  return renderResult;
}
