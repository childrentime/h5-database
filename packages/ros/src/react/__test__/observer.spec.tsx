/**
 * @jest-environment jsdom
 */

import { describe, it, expect, jest, afterEach } from "@jest/globals";
import { observable } from "../../observables";
import { observer } from "../observer";
import { act, cleanup, fireEvent, render } from "@testing-library/react";
import {
  StrictMode,
  createElement,
  createRef,
  forwardRef,
  memo,
  useImperativeHandle,
  useRef,
} from "react";
import { useObserver } from "../useObserver";
import { observable as observableDecorator } from "../../decorator";
import "@testing-library/jest-dom/jest-globals";

afterEach(cleanup);

let consoleWarnMock: any;
afterEach(() => {
  consoleWarnMock?.mockRestore();
});

describe("nest rendering", () => {
  const execute = () => {
    const store = observable.object({
      todos: [
        {
          completed: false,
          title: "a",
        },
      ],
    }).value;
    const renderings = {
      item: 0,
      list: 0,
    };

    const TodoItem = observer(({ todo }: { todo: (typeof store.todos)[0] }) => {
      renderings.item++;
      return <li>|{todo.title}</li>;
    });

    const TodoList = observer(() => {
      renderings.list++;
      return (
        <div>
          <span>{store.todos.length}</span>
          {store.todos.map((todo, idx) => (
            <TodoItem key={idx} todo={todo} />
          ))}
        </div>
      );
    });
    const rendered = render(<TodoList />);
    return { ...rendered, store, renderings };
  };

  it("first rendering", () => {
    const { getAllByText, renderings } = execute();
    expect(renderings.list).toBe(1);
    expect(renderings.item).toBe(1);
    expect(getAllByText("1")).toHaveLength(1);
    expect(getAllByText("|a")).toHaveLength(1);
  });

  it("inner store changed", () => {
    const { store, getAllByText, renderings } = execute();
    act(() => {
      store.todos[0].title += "a";
    });
    expect(renderings.list).toBe(2);
    expect(renderings.item).toBe(2);
    expect(getAllByText("1")).toHaveLength(1);
    expect(getAllByText("|aa")).toHaveLength(1);
  });

  it("rerendering with outer store added", () => {
    const { store, container, getAllByText, renderings } = execute();
    act(() => {
      store.todos.push({
        completed: true,
        title: "b",
      });
    });
    expect(container.querySelectorAll("li").length).toBe(2);
    expect(getAllByText("2")).toHaveLength(1);
    expect(getAllByText("|b")).toHaveLength(1);
    expect(renderings.list).toBe(2);
    expect(renderings.item).toBe(3);
  });

  it("rerendering with outer store pop", () => {
    const { store, container, renderings } = execute();
    let oldTodo;
    act(() => {
      oldTodo = store.todos.pop();
    });
    expect(renderings.list).toBe(2);
    expect(renderings.item).toBe(1);
    expect(container.querySelectorAll("li").length).toBe(0);
  });
});

describe("should only rerender child component", () => {
  const execute = () => {
    const store = observable.object({
      todos: [
        {
          completed: false,
          title: "a",
        },
      ],
    }).value;
    const store1 = observable.object({ name: "abc" }).value;
    const renderings = {
      item: 0,
      list: 0,
    };

    const TodoItem = observer(() => {
      renderings.item++;
      return <li>{store1.name}</li>;
    });

    const TodoList = observer(() => {
      renderings.list++;
      return (
        <div>
          <span>{store.todos.length}</span>
          <TodoItem />
        </div>
      );
    });
    const rendered = render(<TodoList />);
    return { ...rendered, store, renderings, store1 };
  };

  it("first rendering", () => {
    const { getAllByText, renderings } = execute();
    expect(renderings.list).toBe(1);
    expect(renderings.item).toBe(1);
    expect(getAllByText("1")).toHaveLength(1);
    expect(getAllByText("abc")).toHaveLength(1);
  });

  it("inner store changed", () => {
    const { store1, getAllByText, renderings } = execute();
    act(() => {
      store1.name = "dadsa";
    });
    expect(renderings.list).toBe(1);
    expect(renderings.item).toBe(2);
    expect(getAllByText("1")).toHaveLength(1);
    expect(getAllByText("dadsa")).toHaveLength(1);
  });
});

describe("observers", () => {
  it("observer(forwardRef(cmp)) + useImperativeHandle", () => {
    interface IMethods {
      focus(): void;
    }

    interface IProps {
      value: string;
    }

    const FancyInput = observer(
      forwardRef((props: IProps, ref: React.Ref<IMethods>) => {
        const inputRef = useRef<HTMLInputElement>(null);
        useImperativeHandle(
          ref,
          () => ({
            focus: () => {
              inputRef.current!.focus();
            },
          }),
          []
        );
        return <input ref={inputRef} defaultValue={props.value} />;
      })
    );

    const cr = createRef<IMethods>();
    render(<FancyInput ref={cr} value="" />);
    expect(cr).toBeTruthy();
    expect(cr.current).toBeTruthy();
    expect(typeof cr.current!.focus).toBe("function");
  });

  it("useImperativeHandle and forwardRef should work with useObserver", () => {
    consoleWarnMock = jest.spyOn(console, "warn").mockImplementation(() => {});
    interface IMethods {
      focus(): void;
    }

    interface IProps {
      value: string;
    }

    const FancyInput = memo(
      forwardRef((props: IProps, ref: React.Ref<IMethods>) => {
        const inputRef = useRef<HTMLInputElement>(null);
        useImperativeHandle(
          ref,
          () => ({
            focus: () => {
              inputRef.current!.focus();
            },
          }),
          []
        );
        return useObserver(() => {
          return <input ref={inputRef} defaultValue={props.value} />;
        });
      })
    );

    const cr = createRef<IMethods>();
    render(<FancyInput ref={cr} value="" />);
    expect(cr).toBeTruthy();
    expect(cr.current).toBeTruthy();
    expect(typeof cr.current!.focus).toBe("function");
    expect(consoleWarnMock).toMatchSnapshot();
  });

  it("should hoist known statics only", () => {
    function isNumber() {
      return null;
    }

    function MyHipsterComponent() {
      return null;
    }
    MyHipsterComponent.defaultProps = { x: 3 };
    MyHipsterComponent.propTypes = { x: isNumber };
    MyHipsterComponent.randomStaticThing = 3;
    MyHipsterComponent.type = "Nope!";
    MyHipsterComponent.compare = "Nope!";
    MyHipsterComponent.render = "Nope!";

    const wrapped = observer(MyHipsterComponent) as any;
    expect(wrapped.randomStaticThing).toEqual(3);
    expect(wrapped.defaultProps).toEqual({ x: 3 });
    expect(wrapped.propTypes).toEqual({ x: isNumber });
    expect(wrapped.type).toBeInstanceOf(Function); // And not "Nope!"; this is the wrapped component, the property is introduced by memo
    expect(wrapped.compare).toBe(null); // another memo field
    expect(wrapped.render).toBe(undefined);
  });

  it("should inherit original name/displayName #3438", () => {
    function Name() {
      return null;
    }
    Name.displayName = "DisplayName";
    const TestComponent = observer(Name);

    expect((TestComponent as any).type.name).toBe("Name");
    expect((TestComponent as any).type.displayName).toBe("DisplayName");
  });

  it("parent / childs render in the right order", (done) => {
    // See: https://jsfiddle.net/gkaemmer/q1kv7hbL/13/
    const events: string[] = [];

    class User {
      @observableDecorator() accessor name = "User's name";
    }

    class Store {
      @observableDecorator("object") accessor user: User | null = new User();
      public logout() {
        this.user = null;
      }
    }

    const store = new Store();

    function tryLogout() {
      try {
        store.logout();
        expect(true).toBeTruthy();
      } catch (e) {
        // t.fail(e)
      }
    }

    const Parent = observer(() => {
      events.push("parent");
      if (!store.user) {
        return <span>Not logged in.</span>;
      }
      return (
        <div>
          <Child />
          <button onClick={tryLogout}>Logout</button>
        </div>
      );
    });

    const Child = observer(() => {
      events.push("child");
      if (!store.user) {
        return null;
      }
      return <span>Logged in as: {store.user.name}</span>;
    });

    render(<Parent />);

    tryLogout();
    expect(events).toEqual(["parent", "child"]);
    done();
  });

  it("should have overload for props with children", () => {
    interface IProps {
      value: string;
    }
    const TestComponent = observer<IProps>(({ value }) => {
      return null;
    });

    render(<TestComponent value="1" />);

    // this test has no `expect` calls as it verifies whether such component compiles or not
  });

  it("should preserve generic parameters", () => {
    interface IColor {
      name: string;
      css: string;
    }

    interface ITestComponentProps<T> {
      value: T;
      callback: (value: T) => void;
    }
    const TestComponent = observer(
      <T extends unknown>(props: ITestComponentProps<T>) => {
        return null;
      }
    );

    function callbackString(value: string) {
      return;
    }
    function callbackColor(value: IColor) {
      return;
    }

    render(<TestComponent value="1" callback={callbackString} />);
    render(
      <TestComponent
        value={{ name: "red", css: "rgb(255, 0, 0)" }}
        callback={callbackColor}
      />
    );

    // this test has no `expect` calls as it verifies whether such component compiles or not
  });

  it("should preserve generic parameters when forwardRef", () => {
    interface IMethods {
      focus(): void;
    }

    interface IColor {
      name: string;
      css: string;
    }

    interface ITestComponentProps<T> {
      value: T;
      callback: (value: T) => void;
    }
    const TestComponent = observer(
      <T extends unknown>(
        props: ITestComponentProps<T>,
        ref: React.Ref<IMethods>
      ) => {
        return null;
      }
    );

    function callbackString(value: string) {
      return;
    }
    function callbackColor(value: IColor) {
      return;
    }

    render(<TestComponent value="1" callback={callbackString} />);
    render(
      <TestComponent
        value={{ name: "red", css: "rgb(255, 0, 0)" }}
        callback={callbackColor}
      />
    );

    // this test has no `expect` calls as it verifies whether such component compiles or not
  });

  it("should keep original props types", () => {
    interface TestComponentProps {
      a: number;
    }

    function TestComponent({ a }: TestComponentProps): JSX.Element | null {
      return null;
    }

    const ObserverTestComponent = observer(TestComponent);

    const element = createElement(ObserverTestComponent, { a: 1 });
    render(element);

    // this test has no `expect` calls as it verifies whether such component compiles or not
  });

  it("StrictMode", async () => {
    const o = observable.object({ x: 0 }).value;

    const Cmp = observer(() => {
      return o.x;
    });

    const { container, unmount } = render(
      <StrictMode>
        <Cmp />
      </StrictMode>
    );

    expect(container).toHaveTextContent("0");
    act(() => {
      o.x++;
    });
    expect(container).toHaveTextContent("1");
    unmount();
  });
});
