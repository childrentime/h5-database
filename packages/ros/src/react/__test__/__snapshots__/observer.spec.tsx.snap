// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`observers useImperativeHandle and forwardRef should work with useObserver 1`] = `[MockFunction]`;
